ambry-api/src/main/java/com.github.ambry/router/ByteRange.java

```diff
@@ -236,4 +144,183 @@ public int hashCode() {
      */
     OFFSET_RANGE
   }
+
+  /**
+   * A range from a start offset to an end offset.
+   */
+  private static class ClosedRange extends ByteRange {
+    private long startOffset;
+    private long endOffset;
+
+    /**
+     * @see ByteRange#fromOffsetRange(long, long)
+     */
+    private ClosedRange(long startOffset, long endOffset) {
+      this.startOffset = startOffset;
+      this.endOffset = endOffset;
+    }
+
+    @Override
+    public long getStartOffset() {
+      return startOffset;
+    }
+
+    @Override
+    public long getEndOffset() {
+      return endOffset;
+    }
+
+    @Override
+    public long getRangeSize() {
+      return getEndOffset() - getStartOffset() + 1;
+    }
+
+    @Override
+    public ByteRangeType getType() {
+      return ByteRangeType.OFFSET_RANGE;
+    }
+
+    @Override
+    public ByteRange toResolvedByteRange(long totalSize) {
+      if (getStartOffset() >= totalSize) {
+        throw new IllegalArgumentException("Invalid totalSize: " + totalSize + " for range: " + this);
+      }
+      return new ClosedRange(getStartOffset(), Math.min(getEndOffset(), totalSize - 1));
+    }
+
+    @Override
+    public String toString() {
+      return "ByteRange{startOffset=" + startOffset + ", endOffset=" + endOffset + '}';
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      ClosedRange that = (ClosedRange) o;
+      return startOffset == that.startOffset && endOffset == that.endOffset;
+    }
+
+    @Override
+    public int hashCode() {
+      return 31 * Long.hashCode(startOffset) + Long.hashCode(endOffset);
+    }
+  }
+
+  /**
+   * A range from a start offset to the end of an object.
+   */
+  private static class OpenRange extends ByteRange {
```

Comment:
why is `getRangeSize()` not defined for this?

Reply:
Since only the start offset is known for this range type, so you don't know how long the range is without knowing the total object size.